python/pfs/drp/stella/fitPfsFluxReference.py

```diff
@@ -0,0 +1,978 @@
+    sigma = 0.42466090014400953 * fwhm  # 0.424... = 1/sqrt(log(256))
+    spectrum.flux[:] = GaussianKernel1D(width=sigma / dlambda).convolve(spectrum.flux)
+
+    return spectrum
+
+
+def adjustAbsoluteScale(spectrum, fiberConfig):
+    """Multiply a constant to the spectrum
+    so that its integrations will agree to broadband fluxes.
+
+    Because the broadband fluxes (``fiberConfig.fiberFlux``) are affected
+    by the Galactic extinction, ``spectrum`` must also be reddened before
+    the call to this function.
+
+    Parameters
+    ----------
+    spectrum : `pfs.datamodel.PfsSimpleSpectrum`
+        Spectrum.
+    fiberConfig : `pfs.datamodel.pfsConfig.PfsConfig`
+        PfsConfig that contains only a single fiber.
+
+    Returns
+    -------
+    spectrum : `pfs.datamodel.PfsSimpleSpectrum`
+        The same instance as the argument.
+    """
+    fiberFlux = fiberConfig.fiberFlux[0]
+    fiberFluxErr = fiberConfig.fiberFluxErr[0]
+    filterNames = fiberConfig.filterNames[0]
+
+    refFlux = []
+    for filterName in filterNames:
+        flux = FilterCurve(filterName).photometer(spectrum)
+        refFlux.append(flux)
+
+    refFlux = np.asarray(refFlux, dtype=float)
+
+    # minimum chi^2 solution of
+    #   chi^2 = sum((fiberFlux - scale*refFlux) / fiberFluxErr**2)
+    scale = np.sum(fiberFlux * refFlux / fiberFluxErr**2) / np.sum((refFlux / fiberFluxErr)**2)
+
+    spectrum.flux[:] *= scale
+    return spectrum
+
+
+def calculateSpecChiSquare(obsSpectrum, model, radialVelocity, badMask):
+    """Calculate chi square of spectral fitting
+    between a single observed spectrum and a single model spectrum.
+
+    Parameters
+    ----------
+    obsSpectrum : `pfs.datamodel.pfsFiberArraySet.PfsFiberArray`
+        Observed spectrum.
+    model : `pfs.datamodel.pfsSimpleSpectrum.PfsSimpleSpectrum`
+        Model spectrum.
+    radialVelocity : `float`
+        Radial velocity in km/s.
+    badMask : `List[str]`
+        Mask names.
+
+    Returns
+    -------
+    chisq : `float`
+        chi square.
+    """
+    beta = radialVelocity / const.c.to("km/s").value
+    invDoppler = np.sqrt((1.0 - beta) / (1.0 + beta))
+
+    good = (0 == (model.mask & model.flags.get(*(m for m in badMask if m in model.flags))))
+
+    modelFlux = interpolateFlux(
+        model.wavelength[good], model.flux[good],
+        obsSpectrum.wavelength * invDoppler
+    )
+
+    bad = (0 != (obsSpectrum.mask & obsSpectrum.flags.get(*(m for m in badMask if m in obsSpectrum.flags))))
+
+    flux = np.copy(obsSpectrum.flux)
+    flux[bad] = 0.0
+
+    invVar = 1.0 / obsSpectrum.covar[0, :]
+    invVar[bad] = 0.0
+
+    numer = np.sum((flux * modelFlux) * invVar)
+    denom = np.sum(np.square(modelFlux) * invVar)
+    ampli = numer / denom
+    chisq = np.sum(np.square(flux - ampli * modelFlux) * invVar)
+
+    return chisq
+
+
+def promoteSimpleSpectrumToFiberArray(spectrum, snr):
+    """Promote an instance of PfsSimpleSpectrum to PfsFiberArray.
```

Comment:
Why would we want to do this? Does this imply something lacking in the datamodel implementation?

Reply:
I want to do this because I want to pass a model spectrum to `fitContinuum`. The model spectrum is `PfsSimpleSpectrum` because it does not have `covar`. On the other hand, `fitContinuum` expects `SpectrumSet` and `SpectrumSet.fromPfsArm` expects `PfsFiberArraySet`. For this reason, I have to promote `PfsSimpleSpectrum` to `PfsFiberArraySet`. I don't know whether there are other use cases for these type promotions. If there are, it is more natural that `datamodel` should contain the type promotion methods.